Accepts raw ascii code as delimiter in import tool

tinwelint opened the pull request:
to accomodate for use cases where delimiter is a very specific character
or even if there are no quote characters at all, where \0 can be used.

tinwelint committed:
Accepts raw ascii code as delimiter in import tool

to accomodate for use cases where delimiter is a very specific character
or even if there are no quote characters at all, where \0 can be used.

helt commented:
Will this be available in 2.3.x soon? (unable to build in 2.3.1 on Ubuntu myself)

spacecowboy commented on community/import-tool/src/main/java/org/neo4j/tooling/ImportTool.java:
```diff
@@ -689,10 +689,31 @@ public IdType apply( String from )
+            // Parse "raw" ASCII character style characters:
+            // - \123 --> character with id 123
+            // - \t   --> tab character
+            if ( value.startsWith( "\"" ) && value.length() > 0 )
```
I think you meant `value.length() > 1` here.

spacecowboy replied:
Also, do raw values really start with a `"`? I would have thought a `\`? So shouldn't it be `value.startsWith( "\\" )` ?

tinwelint replied:
Two out of two misses with that condition

spacecowboy commented:
My innate _coverage sense_ is telling me there is a missing test for `if ( raw.equals( "t" ) )`. Could you add a test for that?

spacecowboy commented:
@helt it should be included in `2.3.2` when that is released

systay commented on community/import-tool/src/test/java/org/neo4j/tooling/ImportToolTest.java:
```diff
@@ -850,6 +852,40 @@ public void shouldPrintUserFriendlyMessageAboutUnsupportedMultilineFields() thro
+        char weirdDelimiter = 1; // not '1', just the character represented with code 1, which seems to be SOH
+        String name1 = weirdDelimiter + "Weird" + weirdDelimiter;
+        String name2 = "Start " + weirdDelimiter + "middle thing" + weirdDelimiter + " end!";
+        System.out.println( name1 );
```
Forgot to remove?

systay commented:
Should this really be done against 2.2?

tinwelint commented:
My thinking was that it was an innocent change... but perhaps there's no such thing as an innocent change :) . At least 2.3 might have it since it's an issue report for that version.

davidegrohmann commented:
@tinwelint what's the benefit of fixing that problem in 2.3 and not in 2.2?

tinwelint commented:
@davidegrohmann To reduce changes in 2.2

tinwelint commented:
@davidegrohmann @systay @spacecowboy addressed the issues and increased testing on the delimiter converter immensely. There's still the question of version though.

spacecowboy commented on community/import-tool/src/test/java/org/neo4j/tooling/DelimiterConverterTest.java:
```diff
@@ -0,0 +1,110 @@
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+
+public class DelimiterConverterTest
```
IntelliJ reports 100% coverage :+1:

davidegrohmann commented:
@tinwelint my take is if it is a bug than let's fix it in 2.2, if it is a new feature than it should go in 3.0.

spacecowboy commented on community/import-tool/src/test/java/org/neo4j/tooling/ImportToolTest.java:
```diff
@@ -850,6 +852,39 @@ public void shouldPrintUserFriendlyMessageAboutUnsupportedMultilineFields() thro
+        importTool(
+                "--into", dbRule.getStoreDirAbsolutePath(),
+                "--nodes", data.getAbsolutePath(),
+                "--quote", String.valueOf( weirdDelimiter ) );
```
This test does not actually use the delimiter

EDIT: ah it does test as quote.

spacecowboy commented:
I agree that it is fairly unlikely that someone will want to use other characters like `\b` or `\r` as delimiters, so there is no need to support them, but perhaps the IllegalArgumentException should be specific that "Invalid delimiter char '', we only support \t as a special char" or something?

Some missing tests:
- ImportToolTest: Negative test where the character code is invalid
- ImportToolTest: Test character codes for _delimiters_ (as the title of the PR suggests):

```
@Test
    public void shouldAcceptRawAsciiCharacterCodeAsDelimiterConfiguration() throws Exception
    {
        // GIVEN
        List<String> nodeIds = nodeIds();
        Configuration config = new Configuration.Default() {

            @Override
            public char delimiter()
            {
                return '\t';
            }

            @Override
            public char arrayDelimiter()
            {
                return ';';
            }
        };

        // WHEN
        importTool(
                "--into", dbRule.getStoreDirAbsolutePath(),
                "--delimiter", "\\t",
                "--array-delimiter", String.valueOf( config.arrayDelimiter() ),
                "--nodes", nodeData( true, config, nodeIds, alwaysTrue() ).getAbsolutePath(),
                "--relationships",
                                relationshipData( true, config, nodeIds, alwaysTrue(), true ).getAbsolutePath() );

        // THEN
        verifyData();
    }
```

tinwelint commented:
@davidegrohmann it's not a bug, although the current behaviour prevents user from doing an import, which is sort of an issue.

tinwelint commented:
@spacecowboy there are negative tests in DelimiterConverterTest. Now added one in ImportToolTest as well as the one you pasted.

spacecowboy commented:
OK looks good. Waiting on build.
